File: frontend/src/api/http.ts
import { ApiError, type ApiClient } from './client'
import type { ApiErrorBody } from './types'

/**
 * Python バックエンド向けの実装。エンドポイントの一覧は docs/design/04-api.md と対で保つ。
 * 本番の認証は Cloudflare Access(03 §5 の B 案)。Access が付ける JWT をサーバが確かめるので、画面は何も持たない。
 * バックエンドを WORKS_AUTH=dev で動かすときだけ、メールアドレスで入るセッション Cookie になる。
 */
const BASE = '/api/v1'

async function request<T>(method: string, path: string, body?: unknown): Promise<T> {
  const res = await fetch(BASE + path, {
    method,
    credentials: 'same-origin',
    headers: body === undefined ? undefined : { 'Content-Type': 'application/json' },
    body: body === undefined ? undefined : JSON.stringify(body),
  })
  if (res.status === 204) return undefined as T
  const data: unknown = await res.json().catch(() => null)
  if (!res.ok) {
    const err = (data ?? {}) as Partial<ApiErrorBody>
    throw new ApiError(res.status, err.code ?? 'error', err.message ?? `HTTP ${res.status}`)
  }
  return data as T
}

const enc = encodeURIComponent

export function createHttpClient(): ApiClient {
  return {
    // null はログイン画面を出す合図(dev だけ)。Access を通っていない(access_required)・
    // 利用者でない(not_registered)はエラーのまま返し、ログイン画面がその旨を出す
    getSession: () =>
      request<Awaited<ReturnType<ApiClient['getSession']>>>('GET', '/session').catch((e) => {
        if (e instanceof ApiError && e.status === 401 && e.code === 'unauthorized') return null
        throw e
      }),
    login: (email, password) => request('POST', '/session', { email, password }),
    logout: async () => {
      const res = await request<{ logout_url?: string } | undefined>('DELETE', '/session')
      if (res?.logout_url) {
        // Access のセッションを切る。ページごと移るので、呼び出し側の続き(ログイン画面へ送る)は走らせない
        window.location.assign(res.logout_url)
        await new Promise<never>(() => {})
      }
    },

    getMeta: () => request('GET', '/meta'),
    createObject: (input) => request('POST', '/meta/objects', input),
    updateObject: (key, input) => request('PUT', `/meta/objects/${enc(key)}`, input),
    deleteObject: (key) => request('DELETE', `/meta/objects/${enc(key)}`),
    restoreObject: (key) => request('POST', `/meta/objects/${enc(key)}/restore`),
    reorderObjects: (keys) => request('PUT', '/meta/objects/order', { keys }),
    createView: (object, input) => request('POST', '/meta/views', { object, ...input }),
    updateView: (id, input) => request('PUT', `/meta/views/${enc(id)}`, input),
    deleteView: (id) => request('DELETE', `/meta/views/${enc(id)}`),
    restoreView: (id) => request('POST', `/meta/views/${enc(id)}/restore`),
    reorderViews: (object, ids) => request('PUT', '/meta/views/order', { object, ids }),

    // フィルタと並びは入れ子になるので、クエリ文字列ではなく POST の本文で渡す
    listRecords: (object, params = {}) => request('POST', `/objects/${enc(object)}/records/query`, params),
    getRecord: (object, id) => request('GET', `/objects/${enc(object)}/records/${enc(id)}`),
    createRecord: (object, values) => request('POST', `/objects/${enc(object)}/records`, values),
    updateRecord: (object, id, patch) => request('PATCH', `/objects/${enc(object)}/records/${enc(id)}`, patch),
    deleteRecord: (object, id) => request('DELETE', `/objects/${enc(object)}/records/${enc(id)}`),
    restoreRecord: (object, row) => request('POST', `/objects/${enc(object)}/records/${enc(row.id)}/restore`),

    aggregate: (object, params) => request('POST', `/objects/${enc(object)}/aggregate`, params),
    getTimeline: (object, id) => request('GET', `/objects/${enc(object)}/records/${enc(id)}/timeline`),
    importRecords: (object, params) => request('POST', `/objects/${enc(object)}/import`, params),
    listMcpTokens: () => request('GET', '/settings/mcp/tokens'),
    createMcpToken: (name, client) => request('POST', '/settings/mcp/tokens', { name, client }),
    revokeMcpToken: (id) => request('DELETE', `/settings/mcp/tokens/${enc(id)}`),
    listMcpConnections: () => request('GET', '/settings/mcp/connections'),
    revokeMcpConnection: (id) => request('DELETE', `/settings/mcp/connections/${enc(id)}`),
    getOAuthRequest: (id) => request('GET', `/oauth/requests/${enc(id)}`),
    decideOAuthRequest: (id, approve) => request('POST', `/oauth/requests/${enc(id)}`, { approve }),
    listWebForms: () => request('GET', '/settings/forms'),
    createWebForm: (input) => request('POST', '/settings/forms', input),
    updateWebForm: (id, input) => request('PUT', `/settings/forms/${enc(id)}`, input),
    deleteWebForm: (id) => request('DELETE', `/settings/forms/${enc(id)}`),
    rotateWebFormKey: (id) => request('POST', `/settings/forms/${enc(id)}/rotate`),
    submitWebForm: (key, values) => request('POST', `/forms/${enc(key)}`, values),
    googleStatus: () => request('GET', '/google/status'),
    googleConnect: () => request('POST', '/google/connect'),
    googleDisconnect: () => request('DELETE', '/google/connection'),
    listDriveFiles: (q) => request('GET', `/drive/files?q=${enc(q)}`),
    createDriveDocument: (object, id, field) => request('POST', `/objects/${enc(object)}/records/${enc(id)}/drive/${enc(field)}/document`),
    exportRecords: async (object, params = {}) => {
      const res = await fetch(`${BASE}/objects/${enc(object)}/export`, {
        method: 'POST',
        credentials: 'same-origin',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(params),
      })
      if (!res.ok) throw new ApiError(res.status, 'error', `HTTP ${res.status}`)
      return res.blob()
    },
    search: (q) => request('GET', `/search?q=${enc(q)}`),
  }
}

File: backend/app/google/http.py
"""Google の API を叩く**1 本の口**。テストはこの `call` だけを差し替える(本物へは繋がない)。

失敗は全部ここで `ApiError` に訳す。呼ぶ側は HTTP の詳細を知らない。
"""

from typing import Any

import httpx2

from app.errors import ApiError

TIMEOUT = 15.0


def reauth_needed(message: str = "Google の接続が切れました。もう一度繋いでください") -> ApiError:
    """許可が取り消された・鍵が変わった。画面は「Google に接続」を出し直す。"""
    return ApiError(409, "google_reauth", message)


def _message(res: httpx2.Response) -> str:
    """Google のエラーの本文から、人に見せる 1 行を取り出す。"""
    try:
        body = res.json()
    except ValueError:
        return res.text[:200]
    error = body.get("error") if isinstance(body, dict) else None
    if isinstance(error, dict):
        text = error.get("message") or error.get("status") or ""
        return str(text)[:200]
    if isinstance(error, str):
        # OAuth のエラーは {"error": "invalid_grant", "error_description": "..."} の形
        detail = body.get("error_description") or ""
        return f"{error}{f': {detail}' if detail else ''}"[:200]
    return res.text[:200]


def call(
    method: str,
    url: str,
    *,
    headers: dict[str, str] | None = None,
    params: dict[str, Any] | None = None,
    json: dict[str, Any] | None = None,
    data: dict[str, Any] | None = None,
) -> dict[str, Any]:
    try:
        with httpx2.Client(timeout=TIMEOUT) as client:
            res = client.request(method, url, headers=headers, params=params, json=json, data=data)
    except httpx2.HTTPError as exc:  # 名前が引けない・繋がらない・時間切れ
        raise ApiError(
            502, "google_unavailable", "Google に繋がりませんでした。少し待ってからもう一度お試しください"
        ) from exc

    # 401 は鍵切れ、403 の「権限が足りない」は許可の取り消し・スコープ変更。どちらも繋ぎ直しで直る
    if res.status_code == 401 or (res.status_code == 403 and "insufficient" in _message(res).lower()):
        raise reauth_needed()
    if res.status_code >= 400:
        raise ApiError(502, "google_error", f"Google の応答が {res.status_code} でした({_message(res)})")
    if not res.content:
        return {}
    body: Any = res.json()
    return body if isinstance(body, dict) else {"items": body}
